Passivating remote EJB Stub fails when rmicCompatible=true

Starting with Jakarta EE 9 / Jakarta Enterprise Beans 4.0, the rmicCompatible property is always true, and this has uncovered what is a pre-existing issue with passivating/activation a remote reference (stub) for EJB 3.x style interfaces that do not extend java.rmi.Remote.

The problem may be seen in the following FAT project/test:

project : com.ibm.ws.ejbcontainer.remote_fat
test : RemoteTests
methods : testMixPassivateSingleton, testEJBHomeWriteValueDirect_EJBHomeTest

The failure is demonstrated with the following exception:

junit.framework.AssertionFailedError: 2021-07-01-19:54:53:324 The response did not contain "[SUCCESS]".  Full output is:"
ERROR: Caught exception attempting to call test method testMixPassivateSingleton on servlet com.ibm.ws.ejbcontainer.remote.singleton.mix.web.PassivationServlet
jakarta.ejb.EJBException: See nested exception; nested exception is: java.lang.ClassCastException: com.ibm.ws.ejbcontainer.remote.singleton.mix.ejb.PassieSingletonRemote
	at com.ibm.ejs.container.util.ExceptionUtil.EJBException(ExceptionUtil.java:399)
	at com.ibm.ejs.container.util.ExceptionUtil.EJBException(ExceptionUtil.java:285)
	at com.ibm.ejs.container.BusinessExceptionMappingStrategy.mapException(BusinessExceptionMappingStrategy.java:347)
	at com.ibm.ejs.container.BusinessExceptionMappingStrategy.setUncheckedException(BusinessExceptionMappingStrategy.java:519)
	at com.ibm.ejs.container.EJSDeployedSupport.setUncheckedException(EJSDeployedSupport.java:419)
	at com.ibm.ejs.container.EJSContainer.preinvokeHandleException(EJSContainer.java:2652)
	at com.ibm.ejs.container.EJSContainer.EjbPreInvoke(EJSContainer.java:2433)
	at com.ibm.ws.ejbcontainer.remote.singleton.mix.shared.EJSLocal0SFStatefulEJBRefBean_b5dda32b.testRemoteSingleStart(EJSLocal0SFStatefulEJBRefBean_b5dda32b.java)
	at com.ibm.ws.ejbcontainer.remote.singleton.mix.web.PassivationServlet.testMixPassivateSingleton(PassivationServlet.java:51)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at componenttest.app.FATServlet.doGet(FATServlet.java:71)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:500)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:587)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.service(ServletWrapper.java:1258)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:746)
	at com.ibm.ws.webcontainer.servlet.ServletWrapper.handleRequest(ServletWrapper.java:443)
	at com.ibm.ws.webcontainer.filter.WebAppFilterManager.invokeFilters(WebAppFilterManager.java:1227)
	at com.ibm.ws.webcontainer.webapp.WebApp.handleRequest(WebApp.java:5049)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.handleRequest(DynamicVirtualHost.java:314)
	at com.ibm.ws.webcontainer.WebContainer.handleRequest(WebContainer.java:1007)
	at com.ibm.ws.webcontainer.osgi.DynamicVirtualHost$2.run(DynamicVirtualHost.java:279)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink$TaskWrapper.run(HttpDispatcherLink.java:1159)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.wrapHandlerAndExecute(HttpDispatcherLink.java:428)
	at com.ibm.ws.http.dispatcher.internal.channel.HttpDispatcherLink.ready(HttpDispatcherLink.java:387)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleDiscrimination(HttpInboundLink.java:566)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.handleNewRequest(HttpInboundLink.java:500)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.processRequest(HttpInboundLink.java:360)
	at com.ibm.ws.http.channel.internal.inbound.HttpInboundLink.ready(HttpInboundLink.java:327)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.sendToDiscriminators(NewConnectionInitialReadCallback.java:167)
	at com.ibm.ws.tcpchannel.internal.NewConnectionInitialReadCallback.complete(NewConnectionInitialReadCallback.java:75)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.requestComplete(WorkQueueManager.java:504)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.attemptIO(WorkQueueManager.java:574)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager.workerRun(WorkQueueManager.java:958)
	at com.ibm.ws.tcpchannel.internal.WorkQueueManager$Worker.run(WorkQueueManager.java:1047)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:238)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassCastException: com.ibm.ws.ejbcontainer.remote.singleton.mix.ejb.PassieSingletonRemote
	at com.ibm.ws.transport.iiop.internal.WSPortableRemoteObjectImpl.narrow(WSPortableRemoteObjectImpl.java:50)
	at javax.rmi.PortableRemoteObject.narrow(PortableRemoteObject.java:62)
	at com.ibm.ejs.container.SerializedStub.readResolve(SerializedStub.java:41)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.io.ObjectStreamClass.invokeReadResolve(ObjectStreamClass.java:1265)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2277)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1768)
	at java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2541)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2433)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1768)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:548)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:458)
	at java.base/java.util.ArrayList.readObject(ArrayList.java:928)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1175)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2400)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1768)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:548)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:458)
	at java.base/java.util.HashMap.readObject(HashMap.java:1460)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.io.ObjectStreamClass.invokeReadObject(ObjectStreamClass.java:1175)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2400)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1768)
	at java.base/java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2541)
	at java.base/java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2433)
	at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2266)
	at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1768)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:548)
	at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:458)
	at com.ibm.ejs.container.passivator.StatefulPassivator.activate(StatefulPassivator.java:464)
	at com.ibm.ejs.container.StatefulBeanO.activate(StatefulBeanO.java:781)
	at com.ibm.ejs.container.activator.StatefulSessionActivationStrategy.atActivate(StatefulSessionActivationStrategy.java:342)
	at com.ibm.ejs.container.activator.Activator.preInvokeActivateBean(Activator.java:263)
	at com.ibm.ejs.container.EJSContainer.preInvokeActivate(EJSContainer.java:3001)
	at com.ibm.ejs.container.EJSContainer.EjbPreInvoke(EJSContainer.java:2390)
". expected:<0> but was:<1>
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:537)
	at componenttest.topology.utils.HttpUtils.findStringInHttpConnection(HttpUtils.java:495)
	at componenttest.topology.utils.HttpUtils.findStringInUrl(HttpUtils.java:472)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:444)
	at componenttest.topology.utils.HttpUtils.findStringInReadyUrl(HttpUtils.java:414)
	at componenttest.topology.utils.FATServletClient.runTest(FATServletClient.java:83)
	at componenttest.custom.junit.runner.SyntheticServletTest.invokeExplosively(SyntheticServletTest.java:40)
	at componenttest.custom.junit.runner.FATRunner$1.evaluate(FATRunner.java:197)
	at componenttest.rules.repeater.RepeatTests$CompositeRepeatTestActionStatement.evaluate(RepeatTests.java:115)
	at componenttest.custom.junit.runner.FATRunner$2.evaluate(FATRunner.java:318)
	at componenttest.custom.junit.runner.FATRunner.run(FATRunner.java:171)

--- a/dev/com.ibm.ws.ejbcontainer.remote.portable.core/src/com/ibm/ejs/container/SerializedStub.java
+++ b/dev/com.ibm.ws.ejbcontainer.remote.portable.core/src/com/ibm/ejs/container/SerializedStub.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2011 IBM Corporation and others.
+ * Copyright (c) 2011, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -22,22 +22,26 @@ import javax.rmi.PortableRemoteObject;
  * of this object. The ORB will reconnect stubs stored in instance fields of
  * this serialized object.
  */
-public class SerializedStub
-                implements Serializable
-{
+public class SerializedStub implements Serializable {
     private static final long serialVersionUID = 3019532699780090519L;
 
     private final Object ivStub;
-    private final Class ivClass;
+    private final Class<?> ivClass;
 
-    SerializedStub(Object stub, Class klass)
-    {
+    SerializedStub(Object stub, Class<?> klass) {
         ivStub = stub;
         ivClass = klass;
     }
 
-    private Object readResolve()
-    {
+    private Object readResolve() {
         return PortableRemoteObject.narrow(ivStub, ivClass);
+    }
+
+    public Object getStub() {
+        return ivStub;
+    }
+
+    public Class<?> getInterfaceClass() {
+        return ivClass;
     }
 }

--- a/dev/com.ibm.ws.ejbcontainer.remote/bnd.bnd
+++ b/dev/com.ibm.ws.ejbcontainer.remote/bnd.bnd
@@ -1,5 +1,5 @@
 #*******************************************************************************
-# Copyright (c) 2017, 2020 IBM Corporation and others.
+# Copyright (c) 2017, 2021 IBM Corporation and others.
 # All rights reserved. This program and the accompanying materials
 # are made available under the terms of the Eclipse Public License v1.0
 # which accompanies this distribution, and is available at
@@ -47,6 +47,13 @@ Private-Package: \
   com.ibm.ws.ejbcontainer.remote.internal.ORBObjectFactory, \
   com.ibm.ws.ejbcontainer.remote.internal.ORBObjectFactoryInfo, \
   com.ibm.ws.ejbcontainer.remote.internal.StubObjectResolver
+
+Service-Component: \
+  com.ibm.ws.ejbcontainer.classProvider;\
+    implementation:=com.ibm.ws.serialization.DeserializationClassProvider;\
+    provide:=com.ibm.ws.serialization.DeserializationClassProvider;\
+    properties:="service.vendor=IBM,\
+      classes=com.ibm.ws.ejbcontainer.remote.internal.SerializedStubString"
 
 -buildpath: \
 	com.ibm.ws.org.apache.yoko.corba.spec.1.5;version=latest,\

--- a/dev/com.ibm.ws.ejbcontainer.remote/src/com/ibm/ws/ejbcontainer/remote/internal/SerializedStubString.java
+++ b/dev/com.ibm.ws.ejbcontainer.remote/src/com/ibm/ws/ejbcontainer/remote/internal/SerializedStubString.java
@@ -1,0 +1,51 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2021 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.ejbcontainer.remote.internal;
+
+import java.io.Serializable;
+
+/**
+ * A wrapper for a stub that supports standard serialization.
+ *
+ * When a stub is read using read_value, the ORB will not reconnect the stub.
+ * When rmic compatibility is enabled, stubs for interfaces that are not RMI/IDL
+ * abstract interfaces will extend SerializableStub, which has a writeReplace
+ * method to substitute an instance of SerialiedStub, but that instance will
+ * contain a reference to the original stub which serialization will replace
+ * with a reference back to the same SerializedStub instance. The original stub
+ * is lost.
+ *
+ * The serialization service (via SerializedStubHandler) will compensate by replacing
+ * the SerializedStub with an instance of this class, which holds a reference to the
+ * String representation of the stub rather than the original stub. Then when read
+ * back in the stringified stub will be converted back to the original stub and
+ * narrowed in resolveObject(). The ORB will reconnect the stub.
+ */
+public class SerializedStubString implements Serializable {
+
+    private static final long serialVersionUID = 221516676682695674L;
+
+    private final String stub_string;
+    private final Class<?> remote_interface;
+
+    SerializedStubString(String stub, Class<?> remoteInterface) {
+        stub_string = stub;
+        remote_interface = remoteInterface;
+    }
+
+    String getStub() {
+        return stub_string;
+    }
+
+    Class<?> getInterface() {
+        return remote_interface;
+    }
+}

--- a/dev/com.ibm.ws.ejbcontainer.remote/src/com/ibm/ws/ejbcontainer/remote/internal/StubObjectResolver.java
+++ b/dev/com.ibm.ws.ejbcontainer.remote/src/com/ibm/ws/ejbcontainer/remote/internal/StubObjectResolver.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2015 IBM Corporation and others.
+ * Copyright (c) 2015, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -12,8 +12,10 @@ package com.ibm.ws.ejbcontainer.remote.internal;
 
 import java.io.IOException;
 
+import javax.rmi.PortableRemoteObject;
 import javax.rmi.CORBA.Stub;
 
+import org.omg.CORBA.ORB;
 import org.osgi.framework.ServiceReference;
 import org.osgi.service.component.ComponentContext;
 import org.osgi.service.component.annotations.Activate;
@@ -21,12 +23,15 @@ import org.osgi.service.component.annotations.Component;
 import org.osgi.service.component.annotations.Deactivate;
 import org.osgi.service.component.annotations.Reference;
 
+import com.ibm.ejs.container.SerializedStub;
+import com.ibm.websphere.ras.annotation.Sensitive;
 import com.ibm.ws.serialization.DeserializationObjectResolver;
+import com.ibm.ws.serialization.SerializationObjectReplacer;
 import com.ibm.ws.transport.iiop.spi.ClientORBRef;
 import com.ibm.wsspi.kernel.service.utils.AtomicServiceReference;
 
 @Component
-public class StubObjectResolver implements DeserializationObjectResolver {
+public class StubObjectResolver implements SerializationObjectReplacer, DeserializationObjectResolver {
     private static final String REFERENCE_ORB = "orb";
 
     private final AtomicServiceReference<ClientORBRef> orbRef = new AtomicServiceReference<ClientORBRef>(REFERENCE_ORB);
@@ -47,11 +52,46 @@ public class StubObjectResolver implements DeserializationObjectResolver {
     }
 
     @Override
+    public Object replaceObject(@Sensitive Object object) {
+        // -----------------------------------------------------------------------
+        // When rmic compatibility is enabled, stubs for interfaces that are not
+        // RMI/IDL abstract interfaces will extend SerializableStub, which has a
+        // writeReplace method to substitute an instance of SerializedStub.
+        // Unfortunately, SerializedStub holds a reference to the original stub
+        // which serialization will replace with a reference to the same
+        // SerializdStub returned by writeReplace causing a circular reference
+        // that cannot be deserialized.
+        //
+        // To avoid this, the SerializedStub is replaced here with a very similar
+        // SerializedStubString, using the ORB to convert the Stub to a
+        // String that may be serialized, and then the ORB is used again during
+        // resolveObject to convert the String back to the original Stub.
+        // -----------------------------------------------------------------------
+        if (object instanceof SerializedStub) {
+            ORB orb = orbRef.getServiceWithException().getORB();
+            SerializedStub serializedStub = (SerializedStub) object;
+            String stub_string = orb.object_to_string((org.omg.CORBA.Object) serializedStub.getStub());
+            return new SerializedStubString(stub_string, serializedStub.getInterfaceClass());
+        }
+
+        return null;
+    }
+
+    @Override
     public Object resolveObject(Object object) throws IOException {
         if (object instanceof Stub) {
             ((Stub) object).connect(orbRef.getServiceWithException().getORB());
             return object;
         }
+
+        if (object instanceof SerializedStubString) {
+            ORB orb = orbRef.getServiceWithException().getORB();
+            SerializedStubString serializedStub = (SerializedStubString) object;
+            Object stub = orb.string_to_object(serializedStub.getStub());
+            stub = PortableRemoteObject.narrow(stub, serializedStub.getInterface());
+            return stub;
+        }
+
         return null;
     }
 }

--- a/dev/com.ibm.ws.ejbcontainer.remote_fat/test-applications/EJBHomeTest.war/src/com/ibm/ws/ejbcontainer/remote/fat/home/EJBHomeTestServlet.java
+++ b/dev/com.ibm.ws.ejbcontainer.remote_fat/test-applications/EJBHomeTest.war/src/com/ibm/ws/ejbcontainer/remote/fat/home/EJBHomeTestServlet.java
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2014, 2019 IBM Corporation and others.
+ * Copyright (c) 2014, 2021 IBM Corporation and others.
  * All rights reserved. This program and the accompanying materials
  * are made available under the terms of the Eclipse Public License v1.0
  * which accompanies this distribution, and is available at
@@ -43,13 +43,15 @@ import org.omg.CosNaming.NameComponent;
 import org.omg.CosNaming.NamingContext;
 import org.omg.CosNaming.NamingContextHelper;
 
-import componenttest.annotation.SkipForRepeat;
 import componenttest.app.FATServlet;
 
 @WebServlet("/EJBHomeTestServlet")
 @SuppressWarnings("serial")
 public class EJBHomeTestServlet extends FATServlet {
     private static final Logger logger = Logger.getLogger(EJBHomeTestServlet.class.getName());
+    private static final String rmicCompatibleProperty = System.getProperty("com.ibm.websphere.ejbcontainer.rmicCompatible");
+    private static final boolean rmicCompatible = EJB.class.getName().startsWith("jakarta.")
+                                                  || ("all".equalsIgnoreCase(rmicCompatibleProperty) || "values".equalsIgnoreCase(rmicCompatibleProperty));
 
     @EJB(name = "ejb/home")
     private TestEJBHome home;
@@ -114,10 +116,11 @@ public class EJBHomeTestServlet extends FATServlet {
      * values that should use write_value/read_value. The "WAS EJB 3"
      * marshalling actually uses writeAbstractObject/read_abstract_interface,
      * so a tie mismatch will result in an OutOfMemoryError.
+     * The Enterprise Beans 4.0 marshalling is compatible with RMIC and uses
+     * write_value/read_value so a tie mismatch will result in
+     * CORBA.MARSHAL: Illegal valuetype.
      */
     @Test
-    // TODO: Remove Skip when #17757 is fixed
-    @SkipForRepeat(SkipForRepeat.EE9_FEATURES)
     public void testEJBHomeWriteValueDirect_EJBHomeTest() throws Exception {
         List<?> expected = new ArrayList<Object>(Arrays.asList("a", "b", "c"));
         List<?> actual = stubTestWriteValue((Stub) home.create(), expected);
@@ -132,11 +135,21 @@ public class EJBHomeTestServlet extends FATServlet {
             try {
                 try {
                     OutputStream out = (OutputStream) stub._request("testWriteValue", true);
-                    // "WAS EJB 3" marshalling using writeAbstractObject...
-                    Util.writeAbstractObject(out, value);
+                    if (rmicCompatible) {
+                        // Enterprise Beans 4.0 (or rmicCompatible) marshalling using write_value..
+                        out.write_value((Serializable) value, List.class);
+                    } else {
+                        // "WAS EJB 3" marshalling using writeAbstractObject...
+                        Util.writeAbstractObject(out, value);
+                    }
                     in = (InputStream) stub._invoke(out);
-                    // ...and read_abstract_interface.
-                    return (List<?>) in.read_abstract_interface(List.class);
+                    if (rmicCompatible) {
+                        // ...and read_value.
+                        return (List<?>) in.read_value(List.class);
+                    } else {
+                        // ...and read_abstract_interface.
+                        return (List<?>) in.read_abstract_interface(List.class);
+                    }
                 } catch (ApplicationException ex) {
                     in = (InputStream) ex.getInputStream();
                     String id = in.read_string();

--- a/dev/com.ibm.ws.ejbcontainer.remote_fat/test-applications/SingletonWeb.war/src/com/ibm/ws/ejbcontainer/remote/singleton/mix/web/PassivationServlet.java
+++ b/dev/com.ibm.ws.ejbcontainer.remote_fat/test-applications/SingletonWeb.war/src/com/ibm/ws/ejbcontainer/remote/singleton/mix/web/PassivationServlet.java
@@ -23,7 +23,6 @@ import com.ibm.ws.ejbcontainer.remote.singleton.mix.shared.MixHelper;
 import com.ibm.ws.ejbcontainer.remote.singleton.mix.shared.StatefulEJBRefLocal;
 import com.ibm.ws.ejbcontainer.remote.singleton.mix.shared.StatefulEJBRefRemote;
 
-import componenttest.annotation.SkipForRepeat;
 import componenttest.app.FATServlet;
 
 /**
@@ -45,8 +44,6 @@ public class PassivationServlet extends FATServlet {
      * Passivate and activate an SFSB with a reference to a singleton bean's business interface
      */
     @Test
-    // TODO: Remove Skip when #17757 is fixed
-    @SkipForRepeat(SkipForRepeat.EE9_FEATURES)
     public void testMixPassivateSingleton() throws Exception {
         // Create an instance of the bean by looking up the business interface and ensure the bean contains the default state.
         StatefulEJBRefLocal localBean = lookupLocal();

